src/circuit_proof/r1cs.rs

```diff
@@ -0,0 +1,667 @@
+// implements the Bellman ConstraintSystem trait, so we can use that code/logic.
+// (That would require the bellman code to be decoupled from the underlying pairings.)
+
+/// Represents a variable in our constraint system, where the value represents the index.
+#[derive(Clone, Debug)]
+pub struct Variable(usize);
+
+impl Variable {
+    pub fn get_index(&self) -> usize {
+        self.0
+    }
+}
+
+/// Represents a linear combination of some variables multiplied with their scalar coefficients,
+/// plus a scalar. E.g. LC = variable[0]*scalar[0] + variable[1]*scalar[1] + scalar
+pub struct LinearCombination {
+    variables: Vec<(Variable, Scalar)>,
+    constant: Scalar,
+}
+
+impl LinearCombination {
+    // TODO: make constructor with iterators
+    // see FromIterator trait - [(a1, v1), (a2, v2)].iter().collect() (pass in the iterator, collect to get LC)
+    pub fn construct(variables: Vec<(Variable, Scalar)>, constant: Scalar) -> Self {
+        LinearCombination {
+            variables,
+            constant,
+        }
+    }
+
+    // Used to check that variables in the linear combination are valid
+    pub fn check_variables(&self, var_count: usize) -> Result<(), &'static str> {
+        for (var, _) in &self.variables {
+            let index = var.get_index();
+            if index > var_count - 1 {
+                return Err("Invalid variable index");
+            }
+        }
+        Ok(())
+    }
+
+    pub fn get_variables(&self) -> Vec<(Variable, Scalar)> {
+        self.variables.clone()
+    }
+
+    pub fn get_constant(&self) -> Scalar {
+        self.constant.clone()
+    }
+
+    // evaluate the linear combination, given the variable values in var_assignment
+    pub fn eval(&self, var_assignment: &Vec<Scalar>) -> Scalar {
+        let sum_vars: Scalar = self
+            .variables
+            .iter()
+            .map(|(var, scalar)| scalar * var_assignment[var.get_index()])
+            .sum();
+        sum_vars + self.constant
+    }
+}
+
+/// Represents a vector of groups of 3 linear combinations, where a * b = c
+pub struct ConstraintSystem {
+    // a[i] * b[i] = c[i] for all i
+    a: Vec<LinearCombination>,
+    b: Vec<LinearCombination>,
+    c: Vec<LinearCombination>,
+
+    // Assignments of variables
+    var_assignment: Vec<Scalar>,
+}
+
+impl ConstraintSystem {
+    pub fn new() -> Self {
+        ConstraintSystem {
+            a: vec![],
+            b: vec![],
+            c: vec![],
+            var_assignment: vec![],
+        }
+    }
+    // Allocate a variable and do value assignment at the same time
+    pub fn alloc_variable(&mut self, val: Scalar) -> Variable {
+        self.var_assignment.push(val);
+        Variable(self.var_assignment.len() - 1)
+    }
+
+    // Push one set of linear constraints (a, b, c) to the constraint system.
+    // Pushing a, b, c together prevents mismatched constraints.
+    pub fn push_lc(
```

Comment:
Could this be called `constrain` (or something else that makes the verb more specific than `push`?)